EmbeddingBag w/ differentiable per_sample_weights

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18957 EmbeddingBag w/ differentiable per_sample_weights**
* #18800 EmbeddingBag w/ per_sample_weights CUDA fwd + bwd
* #18799 EmbeddingBag w/ per_sample_weights CPU backward
* #18735 EmbeddingBag CPU forward with per_sample_weights.
* #18734 Refactor CPU embedding_bag implementation

Implemented a backwards pass for per_sample_weights on CPU and CUDA.

Test Plan:
- New tests

gh-metadata: pytorch pytorch 18957 gh/zou3519/31/head

Differential Revision: [D14856526](https://our.internmc.facebook.com/intern/diff/D14856526)

zou3519 committed:
EmbeddingBag w/ differentiable per_sample_weights

Implemented a backwards pass for per_sample_weights on CPU and CUDA.

Test Plan:
- New tests

zou3519 commented:
Timing script: https://gist.github.com/zou3519/0b933110d3efefebe7f4a8207ff5c16c
Timing script output: https://gist.github.com/zou3519/e3eda8e162679a9ce2b3021272307802

I don't think I picked a diverse group of values to benchmark, suggestions for other sizes would be good.

TL;DR for output:
1) EmbeddingBag (denoted by optimized) is generally faster than Embedding + scale + reduction
2) There appears to be something with CUDA Sparse embedding bag unrelated to this PR. It might be optimized for different input sizes or just not work that well.

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -461,6 +461,77 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
+  auto offset2bag_data = offset2bag.data<int64_t>();
+
+  // XXX: 64 was arbitrarily chosen. There is probably a sweet spot for this number.
+  #pragma omp parallel for if (num_samples >= 64)
```
Using parallel.h is preferred here. You can pass in the threshold as an option.

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -461,6 +461,77 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
+    auto bag_idx = offset2bag_data[sample_idx];
+    auto embedding_idx = indices_data[sample_idx];
+
+    output_data[sample_idx] = THBlas_dot<scalar_t>(
```
The reason this was used initially was due to overhead of the corresponding aten function. It might be wroth reevaluating whether this is still necessary.

zou3519 replied:
I think calling BLAS's dot function should be faster than me writing a for loop to manually do the dot product. I can't use PyTorch's dot operator for a few other reasons, though.

zou3519 committed:
Update on "EmbeddingBag w/ differentiable per_sample_weights"

EmbeddingBag w/ differentiable per_sample_weights

Implemented a backwards pass for per_sample_weights on CPU and CUDA.

Test Plan:
- New tests

gh-metadata: pytorch pytorch 18957 gh/zou3519/31/head

zou3519 committed:
Update on "EmbeddingBag w/ differentiable per_sample_weights"

EmbeddingBag w/ differentiable per_sample_weights

Implemented a backwards pass for per_sample_weights on CPU and CUDA.

Test Plan:
- New tests

gh-metadata: pytorch pytorch 18957 gh/zou3519/31/head
